build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1065,14 +1065,15 @@ message OutputFile {
 // [Directory][build.bazel.remote.execution.v2.Directory] protos in a
 // single directory Merkle tree, compressed into one message.
 message Tree {
-  // The root directory in the tree.
-  Directory root = 1;
+  reserved 1, 2; // Used for removed fields in an earlier version of the API.
+
+  // The digest of the root directory in the tree.
+  Digest rootDigest = 3;
 
   // All the child directories: the directories referred to by the root and,
-  // recursively, all its children. In order to reconstruct the directory tree,
-  // the client must take the digests of each of the child directories and then
-  // build up a tree starting from the `root`.
-  repeated Directory children = 2;
+  // recursively, all its children. The keys of the map are the hash of the digest
+  // of each directory.
+  map<string, Directory> directories = 4;
```

Comment:
Using the hash string without a complete `Digest` object would be unusual for the REAPI. The map approach may also be incompatible with #134 and #135.

Have you considered a repeated embedded Digest+Directory message? I wouldn't expect the wire overhead to be significant. However, maybe the generated protobuf code would be less efficient and/or convenient. Or do you see any other downsides?

Reply:
Don't think its incompatible with either of those - we're constrained here by the limitations imposed by proto on maps: they would be Digests if I had my druthers. The keys are the keys, as rendered using hex, whether bytes or strings are used to retain them. Using the digest hash only provides a lookup that seems less awkward than the arbitrary hash/size rendering that we use everywhere else, but I'd be on board for that if necessary. The size of the referring digest can be checked easily with the size of the referent directory as a blob independently.

One other downside is repetition - while trees can't contain loops (and it would be difficult to manufacture one anyway), they can have repeated subtrees, which occurs more often than I'd like to think about. It's quite a departure to use a hierarchical message store in any case - I prefer the digest indexing.